### PR TITLE
ENH: implement ExtensionIndex.insert

### PR DESCRIPTION
**Pull request by jbrockmendel**
Sort of.

### CODE REVIEW

File: pandas/core/indexes/extension.py [hunk 1]
```diff
@@ -242,6 +243,10 @@ def repeat(self, repeats, axis=None):
         result = self._data.repeat(repeats, axis=axis)
         return self._shallow_copy(result)
 
+    def insert(self, loc: int, item):
+        # ExtensionIndex subclasses must override Index.insert
```
Comment (WillAyd) [inline pandas/core/indexes/extension.py]: Any reason not to make an abstractmethod?

Comment (jbrockmendel) [reply pandas/core/indexes/extension.py]: good idea, updated

Comment (WillAyd) [reply pandas/core/indexes/extension.py]: So this I believe is ideal but any concerns on backward compat? Otherwise lgtm

Comment (jbrockmendel) [reply pandas/core/indexes/extension.py]: > any concerns on backward compat?

Nothing comes to mind

Comment (WillAyd) [reply pandas/core/indexes/extension.py]: Gothca. This is non-public right?

**Review by WillAyd — APPROVED** [review]

File: pandas/core/indexes/extension.py (hunk 1 above)
Comment (jschendel) [reply pandas/core/indexes/extension.py]: `ExtensionIndex` needs to inherit from `abc.ABC` in order for the abstractmethod decorator to actually work, right?
```python
In [1]: import abc

In [2]: class Foo: 
   ...:     @abc.abstractmethod 
   ...:     def some_meth(self): 
   ...:         pass 
   ...: 

In [3]: class Bar(Foo): 
   ...:     pass 
   ...: 

In [4]: bar = Bar()

In [5]: class Foo2(abc.ABC): 
   ...:     @abc.abstractmethod 
   ...:     def some_meth(self): 
   ...:         pass 
   ...: 

In [6]: class Bar2(Foo2): 
   ...:     pass 
   ...:

In [7]: bar2 = Bar2()
---------------------------------------------------------------------------
TypeError: Can't instantiate abstract class Bar2 with abstract methods some_meth
```

**Review by simonjayhawkins — APPROVED** [review]
lgtm ex https://github.com/pandas-dev/pandas/pull/32476#discussion_r389193483

also, since the type of Index returned depends on the inserted item may be worth adding return types so we get mypy checking consistency.

**Comment by jbrockmendel** [comment]
The abc.abstractmethod didn't appear to function as intended even when I subclassed abc.ABC, so reverted to use AbstractMethodError.

**Comment by jbrockmendel** [comment]
comments addressed, sort of.

**Comment by jreback** [comment]
seems fine

**Pull request merged**